src/oaklib/implementations/sqldb/sql_implementation.py

```diff
@@ -1065,8 +1065,58 @@ def _rbox_relationships(
     def relationships_metadata(
         self, relationships: Iterable[RELATIONSHIP], **kwargs
     ) -> Iterator[Tuple[RELATIONSHIP, List[Tuple[PRED_CURIE, Any]]]]:
-        for rel in relationships:
-            anns = [(ann.predicate, ann.object) for ann in self._axiom_annotations(*rel)]
+        """For given relationships, returns axiom annotation metadata"""
+        # Fetch all annotations
+        anns: List[om.Annotation] = self._axiom_annotations_all()
+        # Simplify objects
+        anns: List[Dict[str, str]] = [{
+            'subject': x.subject,
+            'object': x.object,
+            'predicate': x.predicate,
+            'annotation_predicate': x.annotation_predicate,
+            'annotation_value_or_obj': x.annotation_value if x.annotation_value is not None else x.annotation_object,
+        } for x in anns]
+        # Create lookup for annotations based using edges as keys
+        rel_anns: Dict[RELATIONSHIP, List[Tuple[PRED_CURIE, Any]]] = {}
+        for x in anns:
+            key = (x['subject'], x['predicate'], x['object'])
+            if key not in rel_anns:
+                rel_anns[key] = []
+            rel_anns[key].append((x['annotation_predicate'], x['annotation_value_or_obj']))
+        # Group annotations by relationship, and yield
+        rel_anns = {rel: rel_anns[rel] if rel in rel_anns else [] for rel in relationships}
+        for rel, anns in rel_anns.items():
+            yield rel, anns
+
+    def relationships_metadata2(
+        self, relationships: Iterable[RELATIONSHIP], **kwargs
+    ) -> Iterator[Tuple[RELATIONSHIP, List[Tuple[PRED_CURIE, Any]]]]:
+        """For given relationships, returns axiom annotation metadata"""
+        # Split based on predicate
+        preds: Set[PRED_CURIE] = {x[1] for x in relationships}
```

Comment:
remember the input is an Iterable, so if it's an iterator, not a collection, this will consume it

Reply:
Good catch! I added a line above adding `rels` and using that variable instead for the rest of the function.
```py
rels: List[RELATIONSHIP] = [x for x in relationships]
```